mongosql/src/algebrizer/definitions.rs

```diff
@@ -475,14 +478,105 @@ impl<'a> Algebrizer<'a> {
-            expression,
+            expression: expression.clone(),
             is_add_fields,
             cache: SchemaCache::new(),
         });
-        stage.schema(&self.schema_inference_state())?;
-        Ok(stage)
+        project_stage.schema(&self.schema_inference_state())?;
+        Ok(if is_distinct {
+            let group_stage = Self::build_distinct_values_group_and_project_stage(
+                project_stage,
+                self.scope_level,
+                datasources,
+            );
+            group_stage.schema(&self.schema_inference_state())?;
+            group_stage
+        } else {
+            project_stage
+        })
+    }
+
+    pub fn build_distinct_values_group_and_project_stage(
+        source_stage: mir::Stage,
+        scope_level: u16,
+        datasources: BTreeSet<Key>,
+    ) -> mir::Stage {
+        let mut group_keys = Vec::new();
+        let mut bottom: Option<AliasedExpr> = None;
+        let mut datasource_aliases = Vec::new();
+
+        // Identify data sources and build group keys for deduplication
+        for key in datasources.iter() {
+            if key.scope == scope_level {
+                match &key.datasource {
+                    DatasourceName::Named(name) => {
+                        group_keys.push(OptionallyAliasedExpr::Aliased(AliasedExpr {
+                            alias: name.clone(),
+                            expr: Expression::Reference(ReferenceExpr { key: key.clone() }),
+                        }));
+                        datasource_aliases.push(name.clone());
+                    }
+                    DatasourceName::Bottom => {
+                        bottom = Some(AliasedExpr {
+                            alias: BOTTOM_ALIAS.to_string(),
+                            expr: Expression::Reference(ReferenceExpr { key: key.clone() }),
+                        });
+                    }
+                }
+            }
+        }
+
+        if let Some(ref bot) = bottom {
+            group_keys.push(OptionallyAliasedExpr::Aliased(bot.clone()));
+        }
+
+        let group_stage = mir::Stage::Group(mir::Group {
+            source: Box::new(source_stage),
+            keys: group_keys,
+            aggregations: vec![],
+            cache: Default::default(),
+            scope: scope_level,
+        });
+
+        // Build the projection expressions to map the grouped data back to the original structure
+        let mut project_expressions = BindingTuple::new();
+        for datasource_alias in datasource_aliases {
+            project_expressions.insert(
+                Key::named(datasource_alias.as_str(), scope_level),
+                Expression::FieldAccess(FieldAccess {
+                    expr: Box::new(Expression::Reference(ReferenceExpr {
+                        key: Key::bot(scope_level).clone(),
+                    })),
+                    field: datasource_alias.clone(),
+                    is_nullable: false,
```

Comment:
Setting this to false at the moment, I'll continue working on setting this programmatically.

Reply:
I suppose the only way this could be nullable is if the result set coming into the `SELECT DISTINCT` clause is empty. If it is, inevitably this field will be "missing". However, strictly speaking, the result set would therefore be empty so what does the result set schema even represent at that point? I guess we can't statically know it is empty... I think setting this `true` might be the safest bet since we may never be able to statically determine the appropriate value 🤔 Let's spend more time thinking about this for sure.

Reply:
Thanks for the explanation.  I'll set it to true to take the safer route.